File: forex-server/test/request/CandlesDownloadRequestSpec.scala
package request

import org.scalatest.funspec.AnyFunSpec
import org.scalatest.matchers.should.Matchers
import play.api.libs.json.Json

class CandlesDownloadRequestSpec extends AnyFunSpec with Matchers {
  describe("CandlesDownloadRequest format") {
    it("should serialize to and deserialize from JSON correctly") {
      val candlesDownloadRequest = CandlesDownloadRequest(
        EUR_USD,
        H6,
        Seq(Bid),
        "2024-01-01 00:00:00",
        "2025-01-01 00:00:00"
      )
      val expectedJson = Json.obj(
        "currencyPair" -> "EUR_USD",
        "granularity" -> "H6",
        "pricingComponents" -> Json.arr("B"),
        "fromDate" -> "2024-01-01 00:00:00",
        "toDate" -> "2025-01-01 00:00:00"
      )

      // Serialize
      val json = Json.toJson(candlesDownloadRequest)
      json shouldBe expectedJson

      // Deserialize
      val deserialized = json.validate[CandlesDownloadRequest]
      deserialized.isSuccess shouldBe true
      deserialized.get shouldBe candlesDownloadRequest
    }

    it(
      "should properly serialize and deserialize sequences of price components"
    ) {
      val candlesDownloadRequest = CandlesDownloadRequest(
        GBP_JPY,
        M15,
        Seq(Ask, Mid),
        "2024-01-01 00:00:00",
        "2025-01-01 00:00:00"
      )
      val expectedJson = Json.obj(
        "currencyPair" -> "GBP_JPY",
        "granularity" -> "M15",
        "pricingComponents" -> Json.arr("A", "M"),
        "fromDate" -> "2024-01-01 00:00:00",
        "toDate" -> "2025-01-01 00:00:00"
      )

      // Serialize
      val json = Json.toJson(candlesDownloadRequest)
      json shouldBe expectedJson

      // Deserialize
      val deserialized = json.validate[CandlesDownloadRequest]
      deserialized.isSuccess shouldBe true
      deserialized.get shouldBe candlesDownloadRequest
    }
  }
}

File: forex-frontend/src/app/downloader/request/candles-download-request.spec.ts
import { TestBed } from "@angular/core/testing";
import { CandlesDownloadRequest } from "./candles-download-request";
import { CurrencyPair } from "./currency-pair.enums";
import { Granularity } from "./granularity.enums";
import { PricingComponent } from "./pricing-component.enums";

describe("CandlesDownloadRequest", () => {
    beforeEach(() => {
        TestBed.configureTestingModule({});
    });

    it("should serialize to JSON correctly", () => {
        const candlesDownloadRequest = new CandlesDownloadRequest(
            CurrencyPair.AUD_JPY,
            Granularity.H8,
            [PricingComponent.Bid, PricingComponent.Ask],
            "2025-09-07T00:00:00Z",
            "2025-10-07T00:00:00Z"
        );
        const expectedJson =
            '{"currencyPair":"AUD_JPY","granularity":"H8","pricingComponents":["B","A"],"fromDate":"2025-09-07T00:00:00Z","toDate":"2025-10-07T00:00:00Z"}';
        const jsonifiedRequest = JSON.stringify(candlesDownloadRequest);

        expect(jsonifiedRequest).toBe(expectedJson);
    });
});
